plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/TopQueriesService.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.core.service;
+
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.plugin.insights.rules.model.MetricType;
+import org.opensearch.plugin.insights.rules.model.SearchQueryRecord;
+import org.opensearch.plugin.insights.settings.QueryInsightsSettings;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Service responsible for gathering and storing top N queries
+ * with high latency or resource usage
+ *
+ * @opensearch.internal
+ */
+public class TopQueriesService {
+    private boolean enabled;
+    /**
+     * The metric type to measure top n queries
+     */
+    private final MetricType metricType;
+    private int topNSize;
+    /**
+     * The window size to keep the top n queries
+     */
+    private TimeValue windowSize;
+    /**
+     * The current window start timestamp
+     */
+    private long windowStart;
+    /**
+     * The internal thread-safe store that holds the top n queries insight data
+     */
+    private final PriorityBlockingQueue<SearchQueryRecord> topQueriesStore;
```

Comment:
nit: This doesn't need to be a `PriorityBlockingQueue` anymore, because it's accessed by a single thread.